source/extensions/filters/http/oauth2/oauth_client.cc

```diff
@@ -78,15 +117,27 @@ void OAuth2ClientImpl::onSuccess(const Http::AsyncClient::Request&,
                                  Http::ResponseMessagePtr&& message) {
   in_flight_request_ = nullptr;
 
-  ASSERT(state_ == OAuthState::PendingAccessToken);
+  ASSERT(state_ == OAuthState::PendingAccessToken ||
+         state_ == OAuthState::PendingAccessTokenByRefreshToken);
+  const OAuthState oldState = state_;
   state_ = OAuthState::Idle;
 
   // Check that the auth cluster returned a happy response.
   const auto response_code = message->headers().Status()->value().getStringView();
+
   if (response_code != "200") {
     ENVOY_LOG(debug, "Oauth response code: {}", response_code);
     ENVOY_LOG(debug, "Oauth response body: {}", message->bodyAsString());
-    parent_->sendUnauthorizedResponse();
+    switch (oldState) {
+    case OAuthState::PendingAccessToken:
+      parent_->sendUnauthorizedResponse();
+      break;
+    case OAuthState::PendingAccessTokenByRefreshToken:
+      parent_->onUpdateAccessTokenFailure();
+      break;
+    default:
+      throw EnvoyException("Malformed oauth client state");
```

Comment:
NetworkErrorDoubleCallStateInvalid test covers this exception

Reply:
Probably we can explain more about when this case will happen in a real case (?). Also when this crashes envoy, will that be because of the response from the server or upstream?

Reply:
No, the exception is thrown when the programmer will write a bad code. It doesn't depend from answer of oauth server.